<a href="https://colab.research.google.com/github/aichampionslearn/aichampionslearn/blob/main/eda_15_years_of_power_outage.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# EDA: 15 Years of Power Outage

* Title: 15 Years Of Power Outages

* Objective: Predict attrition of your valuable employees

* Kaggle link: https://www.kaggle.com/autunno/15-years-of-power-outages

* Inspired by: http://insideenergy.org/2014/08/18/data-explore-15-years-of-power-outages/#comment-3862651149

This notebook aims to make an EDA (Exploratory Data Analysis) on 15 years of outage to find out the features' relations and to prepare the ground for a Machine Learning model to predict the cause.

# How  this notebook is organized

1. [Data pre-processing](#1.-Data-pre-processing)
2. [Data analysis](#2.-Data-analysis)

# 1. Data pre-processing

We start by importing all the libraries we're going to use:

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.ensemble import RandomForestClassifier

We now need to import our data:

In [33]:
# dataset = pd.read_csv('../input/Grid_Disruption_00_14_standardized - Grid_Disruption_00_14_standardized.csv')
dataset = pd.read_csv("outagedata.csv")
dataset.head()

Event Description  Year Date Event Began Time Event Began  \
0  Severe Weather - Thunderstorms  2014        6/30/2014          8:00 PM   
1  Severe Weather - Thunderstorms  2014        6/30/2014         11:20 PM   
2  Severe Weather - Thunderstorms  2014        6/30/2014          5:55 PM   
3    Fuel Supply Emergency - Coal  2014        6/27/2014          1:21 PM   
4     Physical Attack - Vandalism  2014        6/24/2014          2:54 PM   

  Date of Restoration Time of Restoration  \
0            7/2/2014             6:30 PM   
1            7/1/2014             5:00 PM   
2            7/1/2014             2:53 AM   
3             Unknown             Unknown   
4           6/24/2014             2:55 PM   

                                Respondent       Geographic Areas NERC Region  \
0                 Exelon Corporation/ComEd               Illinois         RFC   
1  Northern Indiana Public Service Company  North Central Indiana         RFC   
2                              We Energies   Southeast  Wisconsin         MRO   
3                              We Energies              Wisconsin         MRO   
4               Tennessee Valley Authority   Nashville, Tennessee        SERC   

  Demand Loss (MW) Number of Customers Affected                          Tags  
0          Unknown                      420,000  severe weather, thunderstorm  
1          Unknown                      127,000  severe weather, thunderstorm  
2              424                      120,000  severe weather, thunderstorm  
3          Unknown                      Unknown   fuel supply emergency, coal  
4          Unknown                      Unknown           vandalism, physical

In [34]:
print("Number of entries: " + str(len(dataset.index)))

Number of entries: 1652


A quick peek at the data shows us that empty values are defined as "Unknown", which means we should treat them as NULL. To decide what to do with each value, we must first analyze how many empty values each column has:

### Dealing with empty values

We can see many columns have "Unknown", which needs to be cleaned. We need to have special care with our numerical columns. Year is pretty likely pretty clean, I expect that "Number of Customers Affected" is more troublesome:

In [35]:
len(pd.to_numeric(dataset['Year'], 'coerce').dropna().astype(int))

1652

"Year" seems to be perfectly filled, we don't need to worry about it.

In [36]:
len(pd.to_numeric(dataset['Demand Loss (MW)'], 'coerce').dropna().astype(int))

807

Over 700 rows are not numeric on 'Demand Loss (MW)'. It's quite a lot of missing values (almost 50%), we'll have to decide if we want to keep it or not.

In [37]:
len(pd.to_numeric(dataset['Number of Customers Affected'], 'coerce').dropna().astype(int))

222

As we can see above, we have too many non numerical rows for this column (only 222 are correctly filled), it may be best to simply drop it. Let's take a quick look at 'Demand Loss (MW)' first:

In [ ]:
print('Demand Loss (MW)')
dataset.iloc[:, 9]

Besides the usual 'None', 'NaN' and 'Unknown', there are some range values in place (separated by '-'). In other circumstances, I would just remove this column, but since we're likely dropping 'Number of Customers Affected', let's simply remove the empty values:

In [39]:
dataset = dataset.iloc[pd.to_numeric(dataset['Demand Loss (MW)'], 'coerce').dropna().astype(int).index, :]
print(len(dataset.index))

807


Let's take a quick look at 'Number of Customers Affected', to make sure it isn't anything we can fix:

In [43]:
print('Number of Customers Affected')
dataset.iloc[:, 10]

Number of Customers Affected


2             120,000
14             65,000
16             38,500
18                NaN
33             89,000
            ...      
1645    40,000-45,000
1647          100,000
1649           81,000
1650          173,000
1651          133,000
Name: Number of Customers Affected, Length: 807, dtype: object

As we suspected, this data is not in good shape to be used. Besides the usual culprits ("NaN", "Unknown", "None"), we also have some strange choices, such as using "Approx. " and " - " to indicate a possible range of values. With that in mind, let's proceed with the plan to drop it:

In [12]:
dataset = dataset[dataset.columns.difference(['Number of Customers Affected'])]

With that done, we can continue with our data pre-processing and replace 'Unknown' with None on all other columns, so that we can have a better idea of how many empty values we have:

In [44]:
for column in dataset.columns:
    dataset[column].replace('Unknown', None, inplace=True)

<ipython-input-44-59b92e0f1770>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dataset[column].replace('Unknown', None, inplace=True)


In [ ]:
dataset.isnull().any()

Many columns have empty values, lets now check how bad it is:

In [47]:
print("Total number of rows: " + str(len(dataset.index)))
print("Number of empty values:")
for column in dataset.columns:
    print(" * " + column + ": " + str(dataset[column].isnull().sum()))

Total number of rows: 807
Number of empty values:
 * Event Description: 0
 * Year: 0
 * Date Event Began: 0
 * Time Event Began: 4
 * Date of Restoration: 8
 * Time of Restoration: 10
 * Respondent: 0
 * Geographic Areas: 2
 * NERC Region: 0
 * Demand Loss (MW): 0
 * Number of Customers Affected: 56
 * Tags: 1


We now have very few columns left with 'None' values, we can just remove these rows.

In [48]:
dataset = dataset.dropna()

Despite "Event Description" being a very interesting column, we don't need it, since it tells the same story as "Tags" (which is simplified):

In [49]:
dataset = dataset[dataset.columns.difference(['Event Description'])]

We can now check if our data is properly cleaned:

In [53]:
print("Total number of rows: " + str(len(dataset.index)))
print("Number of empty values:")
for column in dataset.columns:
    print(" * " + column + ": " + str(dataset[column].isnull().sum()))

Total number of rows: 739
Number of empty values:
 * Date Event Began: 0
 * Date of Restoration: 1
 * Demand Loss (MW): 0
 * Geographic Areas: 0
 * NERC Region: 0
 * Number of Customers Affected: 0
 * Respondent: 0
 * Tags: 0
 * Time Event Began: 0
 * Time of Restoration: 1
 * Year: 0


It is! The new size of our dataset is 797 (down from 1652). We lost quite a lot of data (a little bit over 50%), but it should be enough for our analysis. Before proceeding, let's take a last quick look at it:

In [ ]:
dataset.head()

In [52]:
dataset.replace('Unknown', np.nan, inplace=True)
dataset.replace('Ongoing', np.nan, inplace=True)

In [54]:
dataset.isnull().any()

Date Event Began                False
Date of Restoration              True
Demand Loss (MW)                False
Geographic Areas                False
NERC Region                     False
Number of Customers Affected    False
Respondent                      False
Tags                            False
Time Event Began                False
Time of Restoration              True
Year                            False
dtype: bool

In [55]:
print("Total number of rows: " + str(len(dataset.index)))
print("Number of empty values:")
for column in dataset.columns:
    print(" * " + column + ": " + str(dataset[column].isnull().sum()))

Total number of rows: 739
Number of empty values:
 * Date Event Began: 0
 * Date of Restoration: 1
 * Demand Loss (MW): 0
 * Geographic Areas: 0
 * NERC Region: 0
 * Number of Customers Affected: 0
 * Respondent: 0
 * Tags: 0
 * Time Event Began: 0
 * Time of Restoration: 1
 * Year: 0


# 2. Data analysis

In [21]:
dataset.to_csv("outagedata_clean.csv")

With our dataset properly cleaned, we can now take a look and see how it's distributed (and how the columns relate to each other). A few interesting plots comes to mind:
 * Year and Tags
 * Demand Loss (MW) and Year
 * Count of causes (Tags)
 * Count of occurrences per Year

It's a little too soon for feature engineering, but we should do a little more work on the Tags column before we proceed if we want to be able to do any meaningfull analysis. Since there are many different values, it could be a good idea to rebrand all severe weather columns with just 'severe weather':

In [22]:
dataset.loc[dataset['Tags'].str.contains('severe weather', case=False), 'Tags'] = 'severe weather'

We can now start the plots:

In [ ]:
dim = (12, 30)
fig, ax = plt.subplots(figsize=dim)
sns.swarmplot(x="Year", y="Tags", ax=ax, data=dataset)

In [ ]:
dim = (40, 10)
fig, ax = plt.subplots(figsize=dim)
demand_plot = sns.lmplot(x="Demand Loss (MW)", y="Year", data=dataset)

for item in demand_plot.get_xticklabels():
    item.set_rotation(45)

In [ ]:
dim = (30, 10)
fig, ax = plt.subplots(figsize=dim)
tag_plot = sns.countplot(x="Tags", ax=ax, data=dataset)

for item in tag_plot.get_xticklabels():
    item.set_rotation(45)

In [ ]:
dim = (20, 10)
fig, ax = plt.subplots(figsize=dim)
sns.countplot(x="Year", ax=ax, data=dataset)

From the plots we made, we can tell a few things about power outages:
 * Outages due to severe weather are common almost every year, and is by far the biggest cause in outages.
 * We had a pretty bad year at 2011, but it eventually got better.
 * As times goes by, demand loss increases, even if we have less outages (which is expected, since there are more people on the grid)

Since there are so many severe weather occurrences, it may not be worth to keep the Tags column; we can't make any meaningful prediction with it, since we don't have the number of times that severe weather doesn't cause an outage.

With that in mind, what is left for us to analyze? Well, we have a few options:
 * Build a regression model to predict demand loss.
 * Build a regression to predict the duration of an outage.
 * Build a classifier to predict what part of the day (e.g. morning, afternoon, night, etc) or month of the year outages are more likely to occur.
 * Find out which respondents solve outages faster.
 * Only use data which has number of customers affected known, and find the relation between this data and the other features.

# ADDITIONAL CODE AND WORK

In [32]:
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, confusion_matrix, r2_score, mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

from sklearn.svm import SVR
from sklearn.linear_model import ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import BayesianRidge
from sklearn.neighbors import KNeighborsRegressor